### PR TITLE
[clangd] Improve filtering logic for undesired proto symbols

### PR DESCRIPTION
**Pull request by kadircet**
This used to filter any names with `_` in them, apart from
enum-constants. Resulting in discrepancies in behavior when we had
fields that have `_` in the name, or for accessors like `set_`, `has_`.

The logic seems to be trying to filter mangled names for nested entries,
so adjusted logic to only do so for top-level decls, while still
preserving some public top-level helpers.

Heuristics are still leaning towards false-negatives, e.g. if a
top-level entity has `_` in its name (`message Foo_Bar {}`), it'll be
filtered, or an enum that prefixes its type name to constants
(`enum Foo { Foo_OK }`).

### CODE REVIEW

**Comment by kadircet** [comment]
https://reviews.llvm.org/D46751 for context from the original patch

**Review by hokein — APPROVED** [review]
Thanks, the logic looks reasonable to me.

File: clang-tools-extra/clangd/index/SymbolCollector.cpp [hunk 1]
```diff
@@ -75,18 +76,61 @@ bool isPrivateProtoDecl(const NamedDecl &ND) {
...
-  // that shouldn't be used (Message_Nested). Ignore them completely.
-  // The nested entities are dangling type aliases, we may want to reconsider
-  // including them in the future.
-  // For enum constants, SOME_ENUM_CONSTANT is not private and should be
-  // indexed. Outer_INNER is private. This heuristic relies on naming style, it
-  // will include OUTER_INNER and exclude some_enum_constant.
-  // FIXME: the heuristic relies on naming style (i.e. no underscore in
-  // user-defined names) and can be improved.
-  return (ND.getKind() != Decl::EnumConstant) || llvm::any_of(Name, islower);
+  // There are some internal helpers like _internal_set_foo();
```
[the remark this answers is not in the document]
Comment (kadircet) [reply clang-tools-extra/clangd/index/SymbolCollector.cpp:79]: I feel like that FIXME might still apply. I think it's rather about passing in a custom filter, similar to `FileFilter`, so I am more inclined towards keeping it

File: clang-tools-extra/clangd/index/SymbolCollector.cpp [hunk 2]
```diff
@@ -75,18 +76,61 @@ bool isPrivateProtoDecl(const NamedDecl &ND) {
...
+  // if it has parent's name as a prefix.
+  // This might go wrong if a nested entity has a name that starts with parent's
+  // name, e.g: enum Foo { Foo_X }.
+  if (llvm::isa<EnumConstantDecl>(&ND)) {
+    auto *DC = llvm::cast<EnumDecl>(ND.getDeclContext());
+    if (!DC || !DC->getIdentifier())
+      return false;
+    auto CtxName = DC->getIdentifier()->getName();
+    return !CtxName.empty() && Name.consume_front(CtxName) &&
+           Name.consume_front("_");
```
[the remark this answers is not in the document]
Comment (kadircet) [reply clang-tools-extra/clangd/index/SymbolCollector.cpp:129]: i'd rather keep it as-is, since `CtxName + "_"` might trigger an extra string allocation

**Review by hokein — APPROVED** [review]
still good.

**Pull request merged**